featurebyte/storage/webdav.py

```diff
@@ -0,0 +1,230 @@
+                pat = re.compile(r"<D:href>(.*?)</D:href>")
+                mat = pat.search(response.text)
+                await response.aclose()
+                if mat:
+                    # if the last character is a slash, it is a directory
+                    if mat.group(1).endswith("/"):
+                        continue
+                    # if the last character is not a slash, it is a file
+                    raise FileExistsError("Remote path cannot be created")
+                raise FileExistsError(
+                    "rclone did not return a correct response whilst creating directory"
+                )
+            else:
+                raise FileExistsError("Unknown error occurred while creating directory")
+
+    async def put(self, local_path: Path, remote_path: Path) -> None:
+        """
+        Upload local file to storage
+
+        Parameters
+        ----------
+        local_path: Path
+            Path to local file to be uploaded
+        remote_path: Path
+            Path of remote file to be stored
+
+        Raises
+        ------
+        FileExistsError
+            File or path already exists on remote path
+        """
+        if self.temp:
+            remote_path = Path("temp").joinpath(remote_path)
+
+        # Create all subdirectories
+        await self.mkdir(remote_path)
+
+        request = self.client.build_request(
+            WebdavHTTPMethods.PROPFIND, url=f"{self.base_url}/{remote_path}"
+        )
+        response = await self.client.send(request)
+        if response.status_code != HTTPStatus.NOT_FOUND:
+            raise FileExistsError("File already exists on remote path")
+
+        with open(local_path, "rb") as file_obj:
+            logger.debug(
+                "Put object to storage",
+                extra={
+                    "object_name": str(remote_path),
+                    "file_path": str(local_path),
+                },
+            )
+            response = await self.client.put(
+                url=f"{self.base_url}/{remote_path}", content=file_obj.read()
+            )
+            await response.aclose()
+
+    async def delete(self, remote_path: Path) -> None:
+        """
+        Delete file in storage
+
+        Parameters
+        ----------
+        remote_path: Path
+            Path of remote file to be deleted
+
+        Raises
+        ------
+        FileNotFoundError
+            Remote file does not exist
+        """
+        if self.temp:
+            remote_path = Path("temp").joinpath(remote_path)
+
+        response = await self.client.delete(url=f"{self.base_url}/{remote_path}")
+        await response.aclose()
+        if response.status_code == HTTPStatus.NOT_FOUND:
+            raise FileNotFoundError("Remote file does not exist")
+        if response.status_code == HTTPStatus.NO_CONTENT:
+            return
+
+        # This should not be reached
+        raise FileNotFoundError("")
```

Comment:
Not sure if this is needed. Would it be better to simply assert `response.status_code == HTTPStatus.NO_CONTENT` above?